design/open/lakefs-kv-execution-plan.md

```diff
@@ -0,0 +1,72 @@
+- migrate from old model to new one without downtime
+
+## Per package changes
+
+For the following packages we should handle to move from 'db' package to 'kv'.
+During development it should enable a configuration flag to switch between using kv and db implementation.
+Any move transition should support migration from old Postgres data to new Postgres data format.
+When a complete data model / functionality completes the transition we can exclude the 'option' to swtich from/to and just use the new format.
+
+- Map the use of transactions in order to provide a valid solution using the kv interface
+- Map any database locking to specific solution using the kv implementation
+- Map models from db to key/value key and value format (ProtoBuf?)
+- Map secondary index used - ex: lookup user by id and email
+
+    pkg/auth
+        - metadata information: installation id, version and etc
+        - authorization information: users, group, policy, credentials and etc
+    pkg/actions
+        - actions information: runs, hooks, status
+    pkg/gateway/multiparts
+        - tracking gateway multipart requests
+
+    pkg/graveler/ref
+        - crud: repository, branch, commits, tags (+log)
+        - branch level locking
+        - lock free commit
+    pkg/graveler/retention
+        - uses graveler/ref to iterate over branches and commit log
+    pkg/graveler/staging
+        - staged k/v storage: key/identity/value used to get/set/drop/drop by prefix/list
+
+    pkg/api
+        - calls migrate as part of setup lakefs
+    pkg/catalog
+        - initialize storage with db and lock db
+    pkg/diagnostics
+        - runs list of queries to collect information on the user's environment
+
+## Global changes
+
+- Transition from db to kb should include a configuration flag that enables the use of 'kv' vs the current 'db' implementation.
+  As complete features move to use kv, and migration plan is implemented, we can remove the flag check for the feature and just enable the new functionality.
+- Migrate from current DDL to kv will be supported only for Postgres and dropped when all features works using kv.
+- Basic dump/restore functionality should be supported for the k/v implementation, how do we migrate the information - discovery
+
+## Testing
+
+- Migration - running the same tests on data that was generated with previous db/model
+- Functionality performance - set of tests that will check that we didn't degrade
+- Migration from old format performance - can't take 24h to switch to new version (downtime?)
+- New locking mechanism functionality
+
+## Milestone
+
+- implement adapter k/v for Postgres
+- feature flag (internal) used for kv development
+- each milestone will include migrate db that will use the new kv format. will support only the part we decided as feature complete and will no longer control by the feature flag.
+- for each step: implement the new models to k/v and migrate code
+- implement adapter k/v for other implementations
```

Comment:
```suggestion
- Implement adapter k/v for Postgres
- Feature flag (internal) used for kv development
- Each milestone will include migrate db that will use the new kv format. will support only the part we decided as feature complete and will no longer be controlled by the feature flag.
- For each step: implement the new models to k/v and migrate code
- Implement adapter k/v for other implementations
```

Reply:
Do we plan on releasing versions while some modules are migrated to KV and some not? If so maybe until we complete the entire system transition to KV we should write the data using both formats as a backup until the transition is complete?

Reply:
@nopcoder You might have missed the comment above? I think it's important to address the issue of data integrity while in transition

Reply:
@N-o-Z didn't go over all the comments yet, :/ still on it, thanks!